### PR TITLE
Add a customized version of the resources module, which will be used in test for env setup/cleaning, fixes https://github.com/Azure/azure-powershell/issues/10076

### PR DESCRIPTION
**Pull request by dolauli**

### CODE REVIEW

**Comment by MiYanni** [comment]
Hi @dolauli. I thought the idea was, to have a custom resources module (checked into azure-powershell generation branch), build it, and then just drop it into the [modules](https://github.com/Azure/autorest.powershell/tree/master/powershell/resources/modules) folder here. It would just work without any changes, other than updating the `test-module.ps1` script to have the same local module loading mechanism that the psm1 files have (with allowed local Accounts module to be loaded). This code:
```powershell
  # Load required Az.Accounts module
  $accountsName = 'Az.Accounts'
  $accountsModule = Get-Module -Name $accountsName
  if(-not $accountsModule) {
    $localAccountsPath = Join-Path $PSScriptRoot 'generated\modules'
    if(Test-Path -Path $localAccountsPath) {
      $localAccounts = Get-ChildItem -Path $localAccountsPath -Recurse -Include 'Az.Accounts.psd1' | Select-Object -Last 1
      if($localAccounts) {
        $accountsModule = Import-Module -Name ($localAccounts.FullName) -Scope Global -PassThru
      }
    }
    if(-not $accountsModule) {
      $hasAdequateVersion = (Get-Module -Name $accountsName -ListAvailable | Where-Object { $_.Version -ge [System.Version]'1.6.0' } | Measure-Object).Count -gt 0
      if($hasAdequateVersion) {
        $accountsModule = Import-Module -Name $accountsName -MinimumVersion 1.6.0 -Scope Global -PassThru
      }
    }
  }
```

**Comment by dolauli** [comment]
> Hi @dolauli. I thought the idea was, to have a custom resources module (checked into azure-powershell generation branch), build it, and then just drop it into the [modules](https://github.com/Azure/autorest.powershell/tree/master/powershell/resources/modules) folder here. It would just work without any changes, other than updating the `test-module.ps1` script to have the same local module loading mechanism that the psm1 files have (with allowed local Accounts module to be loaded). This code:
> 
> ```powershell
>   # Load required Az.Accounts module
>   $accountsName = 'Az.Accounts'
>   $accountsModule = Get-Module -Name $accountsName
>   if(-not $accountsModule) {
>     $localAccountsPath = Join-Path $PSScriptRoot 'generated\modules'
>     if(Test-Path -Path $localAccountsPath) {
>       $localAccounts = Get-ChildItem -Path $localAccountsPath -Recurse -Include 'Az.Accounts.psd1' | Select-Object -Last 1
>       if($localAccounts) {
>         $accountsModule = Import-Module -Name ($localAccounts.FullName) -Scope Global -PassThru
>       }
>     }
>     if(-not $accountsModule) {
>       $hasAdequateVersion = (Get-Module -Name $accountsName -ListAvailable | Where-Object { $_.Version -ge [System.Version]'1.6.0' } | Measure-Object).Count -gt 0
>       if($hasAdequateVersion) {
>         $accountsModule = Import-Module -Name $accountsName -MinimumVersion 1.6.0 -Scope Global -PassThru
>       }
>     }
>   }
> ```

Hi @MiYanni There are some reasons that I do not do that.

1. Auto generated modules depend on Az.Accounts for A&A, but they do not have any dependency on Az.Resources. And Az.Resources is only required when a user tries to run test. So putting Az.Resources into the auto generated modules is not necessary.
2. Putting source code in one repo and then putting the compiled binary in another are really hard to maintain.
3. When we upgrade the compiler, we may need to rebuild the binary and recheck-in if there are compatible issue.
4. Besides, we just want to import the customized version of the resource module, which is different from Az.Resources.

**Comment by MiYanni** [comment]
@dolauli 
1. The modules folder is not 'part of the generated module'. It is put there when generating the module, and is used if Az.Accounts is not installed on the system (great for development purposes). However, it doesn't get shipped with the module. Everything in the 'generated' folder is not packaged with the module.
2. From my understanding, this mini-version of Az.Resources (let's call it Az.TestResources), would basically only need to be built once. The functionality required is minimal and would likely not change for years. If it does need to be changed, it would be in the same repo with all other modules, so it isn't out-of-place.
3. I don't know what this entails. But, you can always build the module with a set version of autorest or autorest.powershell. The NPM package versions don't go away.
4. Agreed. Which is why it should be named appropriately. To me, it just makes sense to keep all modules (even those that aren't put onto the AzGallery) in the same repo. Building this module dynamically in every repo seems a bit overkill.

As long as Az.Accounts stays forward compatible, a static version of Az.TestResources would continue working. However, I'm just advising. Do what you think is best.

**Comment by dolauli** [comment]
@MiYanni Thanks for your comments. Full understand what you suggested. Just feel a little weird to check in binaries into code repo. Besides, regarding to the build time of the mini version resource module, it takes about 20s. And it only happens once when the user first runs test-module.ps1 within a module.
@fearthecowboy @markcowl @erich-wang Call for more suggestions.

File: powershell/resources/assets/tools/Resources/resources/ModelSurface.md [hunk 1]
```diff
@@ -0,0 +1,1645 @@
+### AddOwnerParameters \<Object\> [Api16]
```
Comment (markcowl) [inline powershell/resources/assets/tools/Resources/resources/ModelSurface.md]: No reason to include this file

File: powershell/resources/assets/tools/readme.azure.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
+# Azure PowerShell AutoRest Configuration
+
```
Comment (markcowl) [inline powershell/resources/assets/tools/readme.azure.md]: We can just put this inline with the readme above

File: powershell/resources/assets/tools/Resources/Az.Resources.psd1 [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
...oupMember', 'Remove-AzADGroupOwner', 'Remove-AzADServicePrincipal', 'Remove-AzADServicePrincipalKeyCredential', 'Remove-AzADServicePrincipalPasswordCredential', 'Remove-AzADUser', 'Remove-AzDeployment', 'Remove-AzManagedApplication', 'Remove-AzManagedApplicationDefinition', 'Remove-AzManagementGroup', 'Remove-AzManagementGroupSubscription', 'Remove-AzPolicyAssignment', 'Remove-AzPolicyDefinition', 'Remove-AzPolicySetDefinition', 'Remove-AzResource', 'Remove-AzResourceGroup', 'Remove-AzResourceLock', 'Remove-AzRoleAssignment', 'Remove-AzRoleDefinition', 'Remove-AzTag', 'Restore-AzADDeletedApplication', 'Set-AzManagedApplication', 'Set-AzManagedApplicationDefinition', 'Set-AzManagementGroup', 'Set-AzPolicyDefinition', 'Set-AzPolicySetDefinition', 'Set-AzResource', 'Set-AzResourceGroup', 'Set-AzResourceLock', 'Set-AzRoleDefinition', 'Stop-AzDeployment', 'Test-AzADGroupMember', 'Test-AzDeployment', 'Test-AzNameAvailability', 'Test-AzResourceGroup', 'Test-AzResourceMove', 'Unregister-AzResourceProvider', 'Update-AzADApplication', 'Update-AzADServicePrincipal', 'Update-AzADUser', 'Update-AzManagedApplication', 'Update-AzManagementGroup', 'Update-AzResource', 'Update-AzResourceGroup', '*'
```
[the remark this answers is not in the document]
Comment (dolauli) [reply powershell/resources/assets/tools/Resources/Az.Resources.psd1]: Below are what we actually support based on our previous discussion.
CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Alias           Get-AzResourceGroupDeployment                      0.0.1      Az.Resources.TestSupport
Alias           New-AzResourceGroupDeployment                      0.0.1      Az.Resources.TestSupport
Alias           Test-AzResourceGroupExistence                      0.0.1      Az.Resources.TestSupport
Function        Get-AzDeployment                                   0.0.1      Az.Resources.TestSupport
Function        Get-AzResourceGroup                                0.0.1      Az.Resources.TestSupport
Function        Get-AzResourceProvider                             0.0.1      Az.Resources.TestSupport
Function        New-AzDeployment                                   0.0.1      Az.Resources.TestSupport
Function        New-AzResourceGroup                                0.0.1      Az.Resources.TestSupport
Function        Remove-AzResourceGroup                             0.0.1      Az.Resources.TestSupport
Function        Test-AzResourceGroup                               0.0.1      Az.Resources.TestSupport

File: powershell/resources/assets/tools/Resources/Az.Resources.psd1 [hunk 4]
```diff
@@ -0,0 +1,25 @@
+@{
```
[the remark this answers is not in the document]
Comment (dolauli) [reply powershell/resources/assets/tools/Resources/Az.Resources.psd1]: will make the change

File: powershell/resources/assets/tools/Resources/license.txt [hunk 5]
```diff
@@ -0,0 +1,227 @@
...
+      incurred by, or claims asserted against, such Contributor by reason
+      of your accepting any such warranty or additional liability.
+
+-------------------END OF LICENSE------------------------------------------
+
+
+----------------START OF THIRD PARTY NOTICE--------------------------------
+
+
+The software includes the AutoMapper library ("AutoMapper").  The MIT License set out below is provided for informational purposes only. It is not the license that governs any part of the software.
```
[the remark this answers is not in the document]
Comment (dolauli) [reply powershell/resources/assets/tools/Resources/license.txt]: Will remote it.

**Comment by dolauli** [comment]
In the updated PR, I generates and builds the resource module in a shared location ($HOME/.PSSharedModules/Resources), which will save time from building the module for each modules.

**Pull request merged**